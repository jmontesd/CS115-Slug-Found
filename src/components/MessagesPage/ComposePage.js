import React from 'react';
import './MessagesPage.scss';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';

const MessagesPage = (props) => {
  const { isLoggedIn } = props;

  if (!isLoggedIn) return <Redirect to="/login" />;

  return (
    <div className="container">
      <h1>Send A Private Message</h1>
      <div className="to">To:</div>
      <textarea name="receiver" id="usrname" rows="1" cols="30" />
      <div className="to">Subject:</div>
      <textarea name="subject" rows="1" cols="40" />
      <div className="to">Message:</div>
      <textarea name="message" rows="10" cols="80" />
      <div className="sub">
        <input type="submit" value="Send" />
      </div>
    </div>
  );
};
const mapStateToProps = (state) => {
  return {
    isLoggedIn: state.firebase.auth.uid,
  };
};

export default connect(mapStateToProps)(MessagesPage);
