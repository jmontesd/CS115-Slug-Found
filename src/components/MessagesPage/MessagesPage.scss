.to {
  font-size: 30px;
  padding-bottom: 10px;
  padding-top: 5px;
}

.sub {
  padding-top: 10px;
}

.messages {
  padding: 10px 10px 10px 210px;
  width: 100%;
}

.message {
  clear: both;
  margin-bottom: 5px;
  position: relative;
  padding: 5px;
  background: #f2f2f2;
  border-radius: 5px;
}

.message input {
  position: absolute;
  top: 8px;
  left: 10px;
  width: 30px;
}

.message .sender {
  display: block;
  position: absolute;
  left: 40px;
  color: #000;
  font-weight: bold;
  width: 90px;
}

.message .title {
  display: inline-block;
  padding-left: 150px;
  width: 100%;
}

.message .date {
  position: absolute;
  right: 10px;
}
