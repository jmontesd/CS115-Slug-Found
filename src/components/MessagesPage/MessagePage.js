import React from 'react';
import './MessagesPage.scss';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { compose } from 'redux';
import { firestoreConnect } from 'react-redux-firebase';
import uuid from 'uuid';
import { updateMessages as updateMessagesAction } from '../../store/actions/messageActions';

const MessagesPage = (props) => {
  // necessary props
  const { isLoggedIn, messageGroup, updateMessages, toId } = props;
  // if user not logged in, redirect
  if (!isLoggedIn) return <Redirect to="/login" />;
  // message send handler
  const handleMessageSend = () => {
    updateMessages(toId, 'message', messageGroup);
  };
  // render html
  return (
    <div className="container">
      {messageGroup &&
        messageGroup.messages.map((message) => (
          <div key={uuid()}>{`from${message.from}  message${message.message}`}</div>
        ))}
      <input />
      <button type="button" onClick={handleMessageSend}>
        Send
      </button>
    </div>
  );
};
const mapStateToProps = (state, props) => {
  // get id of to
  const { id: toId } = props.match.params;
  // get all messages
  const { messages: allMessageGroups } = state.firestore.ordered;
  // get id of sender
  const fromId = state.firebase.auth.uid;
  // use to distinguish between message groups
  let smallerId;
  let greaterId;
  if (fromId > toId) {
    greaterId = fromId;
    smallerId = toId;
  } else {
    greaterId = toId;
    smallerId = fromId;
  }
  // find the message group
  const messageGroup =
    allMessageGroups &&
    allMessageGroups.find((m) => m.smallerId === smallerId && m.greaterId === greaterId);

  return {
    messageGroup,
    toId,
    fromId,
    isLoggedIn: state.firebase.auth.uid,
  };
};
// get update message functions
const mapDispatchToProps = (dispatch) => ({
  updateMessages: (toId, message, messageGroup) =>
    dispatch(updateMessagesAction(toId, message, messageGroup)),
});
// connect and export
export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  firestoreConnect([{ collection: 'messages' }]),
)(MessagesPage);
