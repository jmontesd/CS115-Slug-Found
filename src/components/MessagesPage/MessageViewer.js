import React from 'react';
import './MessagesPage.scss';
import uuid from 'uuid';
import moment from 'moment';

export default class MessageViewer extends React.Component {
  state = { message: '' };

  // scroll to bottom once component mounts
  componentDidMount() {
    this.scrollToBottom();
  }

  // scroll to bottom once component updates
  componentDidUpdate() {
    this.scrollToBottom();
  }

  // scroll to bottom function
  scrollToBottom = () => {
    this.messagesEnd.scrollIntoView({ behavior: 'auto' });
  };

  // message send handler
  handleMessageSend = (e) => {
    e.preventDefault();
    const { updateMessages, toId, messageGroup } = this.props;
    const { message } = this.state;
    updateMessages(toId, message, messageGroup);
    this.setState({ message: '' });
  };

  // render message to left or right depending who sent
  // and give color
  renderMessage = (message) => {
    const { toId } = this.props;
    let color = 'orange';
    let direction = 'right';
    if (message.from === toId) {
      color = 'blue';
      direction = 'left';
    }
    return (
      <div className={`message-${color}`} key={uuid()}>
        <p className="message-content">{message.message}</p>
        <div className={`message-timestamp-${direction}`}>
          {moment(message.timestamp).calendar()}
        </div>
      </div>
    );
  };

  render() {
    // necessary props
    const { messageGroup } = this.props;
    const { message } = this.state;
    // render html
    return (
      <>
        <div className="MessageViewer-container">
          {messageGroup && messageGroup.messages.map((m) => this.renderMessage(m))}
          <div
            ref={(e) => {
              this.messagesEnd = e;
            }}
          />
        </div>
        <form
          className="MessageViewer-message-input-and-send"
          onSubmit={(e) => this.handleMessageSend(e)}
        >
          <input
            className="MessageViewer-message-input"
            value={message}
            onChange={(e) => this.setState({ message: e.target.value })}
          />
          <button className="btn btn-secondary MessageViewer-message-button" type="submit">
            Send
          </button>
        </form>
      </>
    );
  }
}
