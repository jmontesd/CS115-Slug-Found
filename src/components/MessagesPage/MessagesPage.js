import React, { useState } from 'react';
import './MessagesPage.scss';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { compose } from 'redux';
import { firestoreConnect } from 'react-redux-firebase';
import uuid from 'uuid';
import MessageViewer from './MessageViewer';
import { updateMessages as updateMessagesAction } from '../../store/actions/messageActions';

export const MessagesPage = (props) => {
  const [messageGroupIndex, setMessageGroupIndex] = useState(0);
  // necessary components
  const { uid, isLoggedIn, users, messageGroups, updateMessages } = props;
  // if user not logged in, redirect
  if (!isLoggedIn) return <Redirect to="/login" />;
  const handleClick = (otherUserId) => {
    messageGroups.forEach((m, index) => {
      if (m.smallerId === otherUserId || m.greaterId === otherUserId) {
        setMessageGroupIndex(index);
      }
    });
  };
  // get the id of user sending message to
  const getToId = () => {
    // make sure there are message groups
    if (messageGroups && messageGroups.length > 0) {
      const { smallerId, greaterId } = messageGroups[messageGroupIndex];
      if (smallerId !== uid) {
        return smallerId;
      }
      return greaterId;
    }
    // there are no message groups
    return -1;
  };
  // render html
  return (
    <div className="container">
      <div className="MessagesPage-container">
        {users.length > 0 ? (
          <>
            <ul className="MessagesPage-container__users list-group list-group-flush">
              {users &&
                users.map(({ user }, index) => (
                  <button
                    type="button"
                    onClick={() => handleClick(user.id)}
                    className={`list-group-item${
                      index === messageGroupIndex ? ' list-group-item-secondary' : ''
                    }`}
                    key={uuid()}
                  >
                    {user && user.username}
                  </button>
                ))}
            </ul>
            <div className="MessagesPage-container__message-page">
              <MessageViewer
                updateMessages={updateMessages}
                messageGroup={
                  messageGroups && messageGroups.length > 0 && messageGroups[messageGroupIndex]
                }
                toId={getToId()}
              />
            </div>
          </>
        ) : (
          <h2>No Messages to show</h2>
        )}
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  // get all messages
  const { messages: allMessageGroups, users: allUsers } = state.firestore.ordered;
  // get userid
  const { uid } = state.firebase.auth;
  // find the message group
  const messageGroups =
    allMessageGroups && allMessageGroups.filter((m) => m.smallerId === uid || m.greaterId === uid);
  // find user
  const users = [];
  if (messageGroups) {
    messageGroups.forEach((messageGroup) => {
      let otherUsersId = messageGroup.smallerId;
      if (otherUsersId === uid) {
        otherUsersId = messageGroup.greaterId;
      }
      // find the user
      const user = allUsers && allUsers.find((i) => i.id === otherUsersId);
      users.push({
        user,
      });
    });
  }
  return {
    uid,
    messageGroups,
    users,
    isLoggedIn: state.firebase.auth.uid,
  };
};
// get update message functions
const mapDispatchToProps = (dispatch) => ({
  updateMessages: (toId, message, messageGroup) =>
    dispatch(updateMessagesAction(toId, message, messageGroup)),
});
// connect and export
export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  firestoreConnect([{ collection: 'messages' }, { collection: 'users' }]),
)(MessagesPage);
