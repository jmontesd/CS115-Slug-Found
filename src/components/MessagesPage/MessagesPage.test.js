import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MessagesPage } from './MessagesPage';

configure({ adapter: new Adapter() });

describe('<MessagesPage />', () => {
  it('should ', () => {
    const wrapper = shallow(<MessagesPage />);
  });
});
