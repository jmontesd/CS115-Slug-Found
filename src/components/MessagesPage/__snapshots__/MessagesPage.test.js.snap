// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessagesPage /> should render MessagesPage properly 1`] = `ShallowWrapper {}`;
